src/borg/item.pyx

```diff
@@ -190,46 +193,59 @@ class PropDict:
         """get value for key, return default if key does not exist"""
         return getattr(self, self._check_key(key), default)
 
-    @staticmethod
-    def _make_property(key, value_type, value_type_name=None, encode=None, decode=None):
-        """return a property that deals with self._dict[key]"""
-        assert isinstance(key, str)
-        if value_type_name is None:
-            value_type_name = value_type.__name__
-        doc = "%s (%s)" % (key, value_type_name)
-        type_error_msg = "%s value must be %s" % (key, value_type_name)
-        attr_error_msg = "attribute %s not found" % key
-
-        def _get(self):
-            try:
-                value = self._dict[key]
-            except KeyError:
-                raise AttributeError(attr_error_msg) from None
-            if decode is not None:
-                value = decode(value)
-            if not isinstance(value, value_type):
-                raise TypeError(type_error_msg)
-            return value
-
-        def _set(self, value):
-            if not isinstance(value, value_type):
-                raise TypeError(type_error_msg)
-            if encode is not None:
-                value = encode(value)
-            self._dict[key] = value
-
-        def _del(self):
-            try:
-                del self._dict[key]
-            except KeyError:
-                raise AttributeError(attr_error_msg) from None
 
-        return property(_get, _set, _del, doc=doc)
+cdef class PropDictProperty:
+    """return a property that deals with self._dict[key] of PropDict"""
+    cdef readonly str key
+    cdef readonly object value_type
+    cdef str value_type_name
+    cdef readonly str __doc__
+    cdef object encode
+    cdef object decode
+    cdef str type_error_msg
+    cdef str attr_error_msg
+
+    def __cinit__(self, value_type, value_type_name=None, encode=None, decode=None):
+       self.key = None
+       self.value_type = value_type
+       self.value_type_name = value_type_name if value_type_name is not None else value_type.__name__
+       self.encode = encode
+       self.decode = decode
+
+    def __get__(self, PropDict instance, owner):
+        try:
+            value = instance._dict[self.key]
+        except KeyError:
+            raise AttributeError(self.attr_error_msg) from None
+        if self.decode is not None:
+            value = self.decode(value)
+        if not isinstance(value, self.value_type):
+            raise TypeError(self.type_error_msg)
```

Comment:
@RonnyPfannschmidt this type check was missing in your PR, I guess you assumed nothing wrong can be in there.

But in `update_internal`, `__set__` is not used, but `_dict` is updated directly.

Reply:
my understanding was that the update_internal method maintains the invariant, because practically - if the data is put into the item, one can no longer get it out if "update_internal" breaks it

Reply:
problem is that when people use `borg transfer`, old data of misc. ages will get read and put into `Items` via `.update_internal(msgpack.unpackb(data)` - maybe quite old stuff that never was really checked for correct type.

so I'ld rather want it to blow up while people beta test this before anything bad gets copied over into borg2 metadata. that why I put quite a lot of type checks all over the place there.

Reply:
i see, lets add a comment about that problem case then, (aka a warning in the update_internam lethod and a comment on the type check that mentions the invariant ignorance of update_internala

Reply:
We're also doing a change from msgpack spec 1 to 2 while going from borg 1.x to 2.0 via borg transfer, so that messes up str and bytes a bit.

But as we know what it should be, that is dealt with by these `want_str` and `want_bytes` (and other) calls.

Reply:
in that case absolutely fabulous